File: snippets/csharp/data-access/read-samples-between.cs
// 
// SQL Race Example: SQL Race Example: Read Samples Between
// 
// Demonstrates: Using GetSamplesBetween to read data in a specific time range
// Prerequisites: MESL.SQLRace.API, a session with data
// Expected output: Sample count and values within the specified range
// Related: read-subsampled-data.cs, reverse-iteration.cs
// 

using MAT.OCS.Core;
using MESL.SqlRace.Domain;

Core.LicenceProgramName = "SQLRace";
Core.Initialize();

var connectionString = $"DbEngine=SQLite;Data Source={Path.Combine(Path.GetTempPath(), "sqlrace-examples.ssn2")};";

if (args.Length == 0)
{
    Console.WriteLine("Usage: dotnet run -- <session-guid> [parameter-identifier]");
    return;
}

SessionKey sessionKey;
try { sessionKey = SessionKey.Parse(args[0]); }
catch (Exception ex) { Console.WriteLine($"ERROR: '{args[0]}' is not a valid GUID. {ex.Message}"); return; }
var paramId = args.Length > 1 ? args[1] : "Temperature:Sensors";

var sessionManager = SessionManager.CreateSessionManager();
IClientSession clientSession;
try { clientSession = sessionManager.Load(sessionKey, connectionString); }
catch (Exception ex)
{
    Console.WriteLine($"ERROR: Could not load session '{sessionKey}': {ex.Message}");
    return;
}
using (clientSession)
{
var session = clientSession.Session;

if (!session.Parameters.Any(p => p.Identifier == paramId))
{
    Console.WriteLine($"ERROR: Parameter '{paramId}' not found in session.");
    Console.WriteLine($"Available: {string.Join(", ", session.Parameters.Take(10).Select(p => p.Identifier))}");
    return;
}

using var pda = session.CreateParameterDataAccess(paramId);

// --- Define a time range: first 500ms of data ---
var rangeStart = session.StartTime;
var rangeEnd = session.StartTime + 500_000_000L; // 0.5 seconds in nanoseconds

var samples = pda.GetSamplesBetween(rangeStart, rangeEnd);

Console.WriteLine($"Parameter: {paramId}");
Console.WriteLine($"Range:     {rangeStart} – {rangeEnd} ns ({(rangeEnd - rangeStart) / 1e9:F3} s)");
Console.WriteLine($"Samples:   {samples.SampleCount}");

for (var i = 0; i < Math.Min(samples.SampleCount, 10); i++)
{
    var relativeMs = (samples.Timestamp[i] - rangeStart) / 1_000_000.0;
    Console.WriteLine($"  t+{relativeMs,8:F3} ms  {samples.Data[i]:F4}");
}

if (samples.SampleCount > 10)
    Console.WriteLine($"  ... and {samples.SampleCount - 10} more");
}

File: snippets/matlab/data-access/read_samples_between.m
% ─────────────────────────────────────────────────────────────
% SQL Race Example: Read Samples Between
% ─────────────────────────────────────────────────────────────
%
% Demonstrates: Using GetSamplesBetween to read data in a specific time range
% Prerequisites: MATLAB R2023a+, .NET 8 runtime, MESL.SqlRace.Domain DLL
%                Run dotnetenv('core') once before using this snippet.
% Input: A .ssn2 SQLite database file (defaults to scenario data)
% Output: Sample count and the first samples within the requested range
%
% Related: snippets/matlab/data-access/reverse_iteration.m
% Docs: https://atlas.motionapplied.com/developer-resources/atlas/sql-race/examples/parameter-data-access/
% ─────────────────────────────────────────────────────────────

% --- Configure .NET runtime (run once per MATLAB session) ---
% dotnetenv('core');  % uncomment if not already set

% --- Load assembly and initialise ---
sqlraceDll = getenv('SQLRACE_DLL_PATH');
if isempty(sqlraceDll)
    sqlraceDll = 'C:\Program Files\McLaren Applied Technologies\ATLAS 10\MESL.SqlRace.Domain.dll';
end

% Pre-load Microsoft.Win32.SystemEvents — required when the runtime config
% does not include Microsoft.WindowsDesktop.App (see docs/troubleshooting.md)
installDir = fileparts(sqlraceDll);
sysEventsDll = fullfile(installDir, 'Microsoft.Win32.SystemEvents.dll');
if exist(sysEventsDll, 'file')
    NET.addAssembly(sysEventsDll);
end

NET.addAssembly(sqlraceDll);
import MESL.SqlRace.Domain.*;

NET.setStaticProperty('MESL.SqlRace.Domain.Core.LicenceProgramName', 'SQLRace');
Core.Initialize();

% --- Connection and session ---
% Default: use the motorsport scenario data file bundled in the repo
dataDir = fullfile(fileparts(mfilename('fullpath')), '..', '..', '..', ...
    'notebooks', 'scenarios', 'data');
dbPath = fullfile(dataDir, 'motorsport-lap-analysis.ssn2');
paramId = 'Temperature:Sensors';

if ~exist(dbPath, 'file')
    error('File not found: %s\nCheck that notebooks/scenarios/data/ contains the .ssn2 files.', dbPath);
end
connectionString = sprintf('DbEngine=SQLite;Data Source=%s;', dbPath);

sessionManager = SessionManager.CreateSessionManager();
sessions = sessionManager.FindBySessionState(SessionState.Historical, connectionString);
if sessions.Count == 0
    error('No sessions found in %s', dbPath);
end
sessionKey = sessions.Item(0).Key;
clientSession = sessionManager.Load(sessionKey, connectionString);

try
    session = clientSession.Session;

    % --- Define a time range: first 500 ms of data ---
    rangeStart = session.StartTime;
    rangeEnd = session.StartTime + int64(500e6);  % 0.5 s in nanoseconds

    pda = session.CreateParameterDataAccess(paramId);
    try
        samples = pda.GetSamplesBetween(rangeStart, rangeEnd);

        fprintf('Parameter: %s\n', paramId);
        fprintf('Range:     %d - %d ns (%.3f s)\n', ...
            rangeStart, rangeEnd, double(rangeEnd - rangeStart) / 1e9);
        fprintf('Samples:   %d\n', samples.SampleCount);

        nShow = min(samples.SampleCount, 10);
        for i = 1:nShow
            relMs = double(samples.Timestamp(i) - rangeStart) / 1e6;
            fprintf('  t+%8.3f ms  %.4f\n', relMs, samples.Data(i));
        end
        if samples.SampleCount > 10
            fprintf('  ... and %d more\n', samples.SampleCount - 10);
        end
    catch ex
        fprintf('Error reading data: %s\n', ex.message);
    end
    pda.Dispose();
catch ex
    fprintf('Error: %s\n', ex.message);
end

clientSession.Dispose();
disp('Done.');
